src/main/java/io/vertx/spi/cluster/impl/zookeeper/ZKAsyncMultiMap.java

```diff
@@ -122,4 +134,42 @@ public void removeAllForValue(V v, Handler<AsyncResult<Void>> completionHandler)
     });
   }
 
+
+  private Map.Entry<String, V> getServerID(ChildData childData) {
+    String[] paths = childData.getPath().split("/");
+    String[] hostAndPort = paths[paths.length - 1].split(":");
+    ServerID serverID = new ServerID(Integer.valueOf(hostAndPort[1]), hostAndPort[0]);
+    Map<String, V> result = new HashMap<>(1);
+    String keyPath = Stream.of(paths).limit(paths.length - 1).reduce((previous, current) -> previous + "/" + current).get();
+    result.put(keyPath, (V) serverID);
+    return result.entrySet().iterator().next();
+  }
+
+  private class Listener implements TreeCacheListener {
+    @Override
+    public void childEvent(CuratorFramework client, TreeCacheEvent treeCacheEvent) throws Exception {
+      ChildData childData;
+      switch (treeCacheEvent.getType()) {
+        case NODE_ADDED:
+          childData = treeCacheEvent.getData();
+          if (childData != null && childData.getData() != null && childData.getData().length > 0) {
+            Map.Entry<String, V> entry = getServerID(childData);
+            ChoosableSet<V> entries = Optional.ofNullable(cache.get(entry.getKey())).orElse(new ChoosableSet<>(1));
+            entries.add(entry.getValue());
+            cache.put(entry.getKey(), entries);
+          }
+          break;
+        case NODE_REMOVED:
+          childData = treeCacheEvent.getData();
+          if (childData != null && childData.getPath() != null) {
```

Comment:
`childData.getPath()` is never going to be null if you follow the checks on validatePath[1].

You probably want the same childData guard used in the NODE_ADDED since `NODE_REMOVED` will be trigged when the last consumer unregisters with a path of `/asyncMultiMap/__vertx.subs/<address>` which then causes the array handling in `getServerID` to die (since it's expecting `/asyncMultiMap/__vertx.subs/<address>/<host>:<port>`).


[1] https://github.com/apache/curator/blob/master/curator-recipes/src/main/java/org/apache/curator/framework/recipes/cache/ChildData.java#L33